src/Algebra/Bundles/Free/Magma.agda

```diff
@@ -0,0 +1,262 @@
+
+  map-id : ∀ (t : PreFreeMagma A) → map id t ≡ t
+  map-id (var a) = ≡-refl
+  map-id (s ∙ t) = cong₂ _∙_ (map-id s) (map-id t)
+
+  map-∘ : (g : A → B) → (f : B → C) → ∀ t → map (f ∘ g) t ≡ (map f ∘ map g) t
+  map-∘ g f (var a) = ≡-refl
+  map-∘ g f (s ∙ t) = cong₂ _∙_ (map-∘ g f s) (map-∘ g f t)
+
+------------------------------------------------------------------------
+-- Functor, RawMonad instance: TODO
+
+------------------------------------------------------------------------
+-- parametrised 'equational' theory over the 'pre'-free algebra
+
+module PreFreeTheory {c ℓ} {A : Set c} (R : Rel A ℓ) where
+
+  data PreFreeMagmaTheory : Rel (PreFreeMagma A) (c ⊔ ℓ)
+
+  open Defs PreFreeMagmaTheory
+
+  data PreFreeMagmaTheory where
+
+    var : ∀ {a b} → (R a b) → PreFreeMagmaTheory (var a) (var b)
+    _∙_ : Congruent₂ _∙_
+
+PreFreeTheorySyntax : ∀ {c ℓ} {A : Set c} (R : Rel A ℓ) → Rel (PreFreeMagma A) (c ⊔ ℓ)
+PreFreeTheorySyntax R = PreFreeMagmaTheory where open PreFreeTheory R
+
+syntax PreFreeTheorySyntax R m n = m ≈[ R ] n
+
+
+------------------------------------------------------------------------
+-- Free algebra on a Set
+{-
+   in the propositional case, we can immediately define the following
+   but how best to organise this under the Algebra.Bundles.Free hierarchy?
+   e.g. should we distinguish Free.Magma.Setoid from Free.Magma.Propositional?
+-}
+
+module FreeMagma {c} (A : Set c) where
+
+  private Carrier = PreFreeMagma A
+
+  _≈_ : Rel Carrier c
+  m ≈ n = m ≈[ _≡_ ] n
+
+  open PreFreeTheory {A = A} _≡_
+
+  ≈⇒≡ : ∀ {m n} → m ≈ n → m ≡ n
+  ≈⇒≡ (var ≡-refl) = ≡-refl
+  ≈⇒≡ (eq₁ ∙ eq₂) = cong₂ _∙_ (≈⇒≡ eq₁) (≈⇒≡ eq₂)
+
+  refl : Reflexive _≈_
+  refl {var _} = var ≡-refl
+  refl {_ ∙ _} = refl ∙ refl
+
+  ≡⇒≈ : ∀ {m n} → m ≡ n → m ≈ n
+  ≡⇒≈ ≡-refl = refl
+
+  rawFreeMagma : RawMagma c c
+  rawFreeMagma = record { Carrier = Carrier ; _≈_ = _≡_ ; _∙_ = _∙_ }
+
+  open Strs {A = Carrier} _≡_
+
+  isMagma : IsMagma _∙_
+  isMagma = record { isEquivalence = ≡-isEquivalence ; ∙-cong = cong₂ _∙_ }
+
+  freeMagma : Magma c c
+  freeMagma = record { RawMagma rawFreeMagma ; isMagma = isMagma }
+
+------------------------------------------------------------------------
+-- Extending to a Setoid
+
+module PreservesEquivalence {c ℓ} {A : Set c} (R : Rel A ℓ) where
+
+  open PreFreeTheory R
+
+  _≈R_ = λ m n → m ≈[ R ] n
+
+  refl : Reflexive R → Reflexive _≈R_
+  refl r {var _} = var r
+  refl r {_ ∙ _} = (refl r) ∙ (refl r)
+
+  sym : Symmetric R → Symmetric _≈R_
+  sym s (var r)   = var (s r)
+  sym s (r₁ ∙ r₂) = sym s r₁ ∙ sym s r₂
+
+  trans : Transitive R → Transitive _≈R_
+  trans t (var r)   (var s)   = var (t r s)
+  trans t (r₁ ∙ r₂) (s₁ ∙ s₂) = trans t r₁ s₁ ∙ trans t r₂ s₂
+
+  preservesEquivalence : IsEquivalence R → IsEquivalence _≈R_
+  preservesEquivalence isEq = record { refl = refl r ; sym = sym s ; trans = trans t }
+    where open IsEquivalence isEq renaming (refl to r; sym to s; trans to t)
```

Comment:
`where module REq = IsEquivalence isEq` and then use `REq.refl` in the record (etc)?

Reply:
Interesting idiom. Yes, perhaps that is better. In any case, all this should now, I think, be folded into the module defining the equations theory.